### PR TITLE
Make the min/max proto functions do something sensible for QUIC

### PR DESCRIPTION
**Pull request by mattcaswell**
Ensure we can set OSSL_QUICV1_VERSION via the min/man protocol API functions. This also applies to the config file equivalents.

### CODE REVIEW

File: doc/man3/SSL_CONF_cmd.pod [hunk 1]
```diff
@@ -416,24 +416,26 @@ This is a synonym for the "Groups" command.
...
 Currently supported protocol values are B<SSLv3>, B<TLSv1>, B<TLSv1.1>,
-B<TLSv1.2>, B<TLSv1.3>, B<DTLSv1> and B<DTLSv1.2>.
-The SSL and TLS bounds apply only to TLS-based contexts, while the DTLS bounds
-apply only to DTLS-based contexts.
-The command can be repeated with one instance setting a TLS bound, and the
-other setting a DTLS bound.
-The value B<None> applies to both types of contexts and disables the limits.
+B<TLSv1.2>, B<TLSv1.3>, B<DTLSv1>, B<DTLSv1.2> and B<QUICv1>.
+The SSL and TLS bounds apply only to TLS-based contexts. The DTLS bounds
+apply only to DTLS-based contexts.The QUIC bounds apply only to QUIC-based
```
Comment (paulidale) [inline doc/man3/SSL_CONF_cmd.pod:421]: Space after full stop.

File: ssl/statem/statem_lib.c [hunk 2]
```diff
@@ -2049,10 +2049,11 @@ int ssl_set_version_bound(int method_version, int version, int *bound)
 
     valid_tls = version >= SSL3_VERSION && version <= TLS_MAX_VERSION_INTERNAL;
     valid_dtls =
-        DTLS_VERSION_LE(version, DTLS_MAX_VERSION_INTERNAL) &&
-        DTLS_VERSION_GE(version, DTLS1_BAD_VER);
+        (version == DTLS1_BAD_VER)
```
Comment (paulidale) [inline ssl/statem/statem_lib.c:2052]: Is this correct?  If it is, there should be a comment explaining what this is a valid DTLS version.

**Review by paulidale — CHANGES_REQUESTED** [review]

File: ssl/statem/statem_lib.c (hunk 2 above)
Comment (mattcaswell) [reply ssl/statem/statem_lib.c:2053]: Yes, it is correct. Comment added.

**Comment by mattcaswell** [comment]
Fixups pushed addressing the review comments. Please take another look.

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
I don't believe we should be using the existing `SSL_set_version` APIs for QUIC version management. Currently these are for forwarding to the handshake layer. How else is the user supposed to be able to configure what version of TLS they want to use for the handshake layer? Currently there is only one version of TLS usable with QUIC but this won't be the case forever.

I would rather introduce a new API for managing the QUIC version.

**Comment by mattcaswell** [comment]
> I don't believe we should be using the existing SSL_set_version APIs for QUIC version management. 

I assume you the "set_min_proto_version" and "set_max_proto_version" APIs (there is no `SSL_set_version`).

I don't agree with you. To me that "set_min_proto_version"/"set_max_proto_version" is not TLS specific at all. It defines the constraints that the protocol version to be negotiated needs to satisfy. For a QUIC object this makes sense that it should be the QUIC protocol version. Of course for QUIC the situation is slightly confused because there are *2* protocol versions to consider - the TLS protocol version and the QUIC protocol version. But in my mind the "primary" protocol version for the object is the QUIC version. The TLS protocol version for QUIC cannot be solely under the configuration of the user since the QUIC protocol version imposes additional constraints on the allowed TLS protocol version. So, for QUICv1 it imposes the constraint that the TLS protocol version must be TLSv1.3.

If in some future version of QUIC other TLS versions are allowed then we can add additional APIs to control the TLS protocol selection at that point. But that is a problem for the future IMO.

> Worth noting, if the motivation here is to reduce application code changes, this actually makes things worse since it means applications have to change or remove existing calls to set a minimum version.

This makes no sense. It seems unreasonable for any application to expect QUIC just to work without any configuration changes at all. So, I don't see that as an objective. However in practice this doesn't matter since, after the change in this PR, if you attempt to set TLS protocol version bounds on a QUIC object then the calls will *succeed* and be ignored. So the outcome is that no configuration changes would be required anyway.

**Comment by hlandau** [comment]
>The TLS protocol version for QUIC cannot be solely under the configuration of the user since the QUIC protocol version imposes additional constraints on the allowed TLS protocol version. So, for QUICv1 it imposes the constraint that the TLS protocol version must be TLSv1.3.

This doesn't follow at all. The mere fact that QUIC requires TLS 1.3 or later does **not** imply that QUICv1 can only be used with TLS 1.3.

>If in some future version of QUIC other TLS versions are allowed then we can add additional APIs to control the TLS protocol selection at that point. But that is a problem for the future IMO.

How is that supposed to work? By this logic we then end up with two APIs to configure TLS versions, one which needs to be used in one circumstance and one which needs to be used in another, which is just confusing to no gain.

The values accepted by `SSL_set_min_proto_version` are also literally TLS version constants from the TLS RFCs. All existing API usage of these functions is in terms of configuring TLS versions. All you do by making this different for QUIC is break existing  code and increase the changes needed in applications, which is the opposite of what we're trying to do.

>This makes no sense. It seems unreasonable for any application to expect QUIC just to work without any configuration changes at all. So, I don't see that as an objective. 

The object was to make an API for single stream QUIC which allows adaptation of existing applications with only minimal changes. The DDD demos and corresponding diffs were done in furtherance of this.

Existing applications will have configuration mechanisms they have developed for configuring a handshake layer. These applications will be calling  `SSL_set_min_proto_version` with a TLS version. I see no reason not to define `SSL_set_min_proto_version` as a handshake-layer function, in which case this all just works automatically. Moreover this proposal leaves no apparent API for the handshake layer protocol version to be configured, even.

It's unclear to me what the gain is supposed to be from this proposed design.

>However in practice this doesn't matter since, after the change in this PR, if you attempt to set TLS protocol version bounds on a QUIC object then the calls will succeed and be ignored. So the outcome is that no configuration changes would be required anyway.

How is `SSL_get_min_proto_version` supposed to work then? If an application calls that it gets an unexpected value. There must be consistency between `SSL_set_min_proto_version` and `SSL_get_min_proto_version`.

**Comment by hlandau** [comment]
RFC 9001
>For brevity, the acronym TLS is used to refer to TLS 1.3, though a newer version could be used; see [Section 4.2](https://www.rfc-editor.org/rfc/rfc9001.html#tls-version).

...

>This document describes how TLS 1.3 [[TLS13](https://www.rfc-editor.org/rfc/rfc9001.html#TLS13)] is used with QUIC.
>
>In practice, the TLS handshake will negotiate a version of TLS to use. This could result in a version of TLS newer than 1.3 being negotiated if both endpoints support that version. This is acceptable provided that the features of TLS 1.3 that are used by QUIC are supported by the newer version.
>
>Clients MUST NOT offer TLS versions older than 1.3. A badly configured TLS implementation could negotiate TLS 1.2 or another older version of TLS. An endpoint MUST terminate the connection if a version of TLS older than 1.3 is negotiated.

**Comment by t8m** [comment]
**OTC: Should these functions affect the QUIC version or the TLS handshake version?**

**Comment by hlandau** [comment]
There are two options being articulated here.

**Option A:** `SSL_set_min_proto_version` (etc.) control the handshake layer version in all cases. This is consistent with all existing usage of these APIs. We introduce some API like `SSL_set_min_quic_version` (etc.) to control the QUIC version (but since we only support QUICv1 for now, we needn't do this immediately.)

An application which does something like:
```c
{
    SSL_set_min_proto_version(s, TLS1_3);
}
```
which is obviously a prolific usage, doesn't need to make any changes to this particular code.

**Option B:** `SSL_set_min_proto_version` (etc.) controls the QUIC version for QUIC, but the TLS version for TLS. This forces us to create some API `SSL_set_min_handshake_proto_version` (etc.) to control the TLS version.

An application would need to change the above code to:
```c
{
   if (QUIC)
        SSL_set_min_handshake_proto_version(s, TLS1_3);
   else
        SSL_set_min_proto_version(s, TLS1_3);
}
```

In either option there will need to be a set of functions for configuring the TLS version and a set of functions for the QUIC version. But Option B essentially flips around these functions when QUIC is used, so the function you need to call to configure the TLS version depends on if you're using QUIC or not. I don't see the benefit here.

**Note on versioning.** It's also worth noting that I am not fully convinced that a `min_proto_version` / `max_proto_version`-style API is going to be entirely suitable for QUIC. It is not necessarily the case that QUIC versions will be allocated monotonically (in terms of their numerical codepoints) by a single body. This means it's entirely plausible in future that an application will want to specify a list of acceptable versions which are not contiguous. So the existing paradigm of version configuration API (re: Option B) is I suspect not well aligned anyway.

Note that the IANA QUIC version numbers registry is a 32-bit space, and has an Expert Review/Specification Required IANA allocation policy (*not* Standards Action). Anyone can create another version of QUIC.

**Comment by davidben** [comment]
@hlandau is correct. This PR isn't right and will cause you all problems down the line.

This PR assumes QUIC is like DTLS in being a sibling to TLS. That's not right. QUIC is layered _on top of_ TLS. In particular, you should _not_ assume the QUIC version uniquely determines the TLS version. If TLS 1.4 gets defined, the default expectation will be that QUICv1 works with it.

(Indeed forcing new TLS versions to become new QUIC versions would make downgrade protection a little tricky, due to how QUIC tends to be signaled out-of-band vs. TLS's in-band signaling.)

**Comment by t8m** [comment]
> This PR assumes QUIC is like DTLS in being a sibling to TLS. That's not right. QUIC is layered _on top of_ TLS. In particular, you should _not_ assume the QUIC version uniquely determines the TLS version. If TLS 1.4 gets defined, the default expectation will be that QUICv1 works with it.

I do not see how anything in this PR implies what you've wrote here. And there is nothing in this PR that would force QUIC version to be tied with TLS versions.

This PR simply assumes that the SSL_get/set_min/max_proto_version() is there for setting the version of the top level protocol (TLS in case of TLS, QUIC in case of QUIC). And yeah, there might be a need to add an extra API for setting the TLS handshake min/max versions at some point when TLS-1.4 is added but we do NOT need to have such call now as there is currently no way in which such call would be useful currently.

**Comment by mattcaswell** [comment]
> I think we need to discuss this within OTC.

Agreed.

> This doesn't follow at all. The mere fact that QUIC requires TLS 1.3 or later does not imply that QUICv1 can only be used with TLS 1.3.

It does for OpenSSL 3.2. There is no implementation of TLS > 1.3 in 3.2.

> The values accepted by SSL_set_min_proto_version are also literally TLS version constants from the TLS RFCs. All existing API usage of these functions is in terms of configuring TLS versions. All you do by making this different for QUIC is break existing code and increase the changes needed in applications, which is the opposite of what we're trying to do.

I just don't agree that this is the case. I consider it bad to do it the other way around.

These functions are commonly invoked via system config file, e.g.

````
system_default = tls_system_default

[tls_system_default]
MaxProtocol = TLSv1.2
MaxProtocol = DTLSv1.2
````

In such a case we should not assume that the system default TLS settings should apply to QUIC. It makes no sense to do so.

> How is SSL_get_min_proto_version supposed to work then? If an application calls that it gets an unexpected value. There must be consistency between SSL_set_min_proto_version and SSL_get_min_proto_version.

Call SSL_get_min_proto_version after a call to SSL_set_min_proto_version with an version that does not apply to that protocol results in a 0 return. Whether you agree with that design or not - these functions *already work that way* (e.g. this is what happens if you use a DTLS version number to set the min protocol version for a TLS based SSL object.

> you should not assume the QUIC version uniquely determines the TLS version.

I am not assuming that. Only that for OpenSSL 3.2 it implies a single TLS version. There is only one supported TLS version supported by 3.2 - TLSv1.3.

In the future if TLSv1.4 gets defined then we will need a new mechanism for defining the allowed TLS protocol versions that applies within a QUIC context. We will need this anyway one way or another (either we need new functions to control the min/max QUIC version in the event of new QUIC versions, or we need new functions to define the allowed TLS versions in the event of TLSv1.4 coming along).

What this comes down to is a debate about whether SSL_set_min/max_proto_version applies to the primary protocol associated with the SSL object (i.e. either QUIC or TLS). Or whether it always applies to the underlying TLS protocol. In my view the former is the way to go and the latter introduces unnecessary differences in the API for QUIC vs TLS (i.e. you are forced to configure the allowed protocol versions differently between QUIC and TLS which goes against what we have been trying to do).

**Comment by hlandau** [comment]
I don't see why we should need or want to have two different functions for configuring which version of TLS to use depending on if QUIC is in use or not. As I note above  `SSL_set_min_proto_version` (etc.) do not even work for QUIC version numbers since QUIC version numbers are not a contiguous sequence anyway.

We can clarify the existing `SSL_set_min_proto_version` function as defining the TLS handshake layer version regardless of whether QUIC is in use or not. This clarification is compatible with all existing usage and with how applications actually use this API.

>These functions are commonly invoked via system config file, e.g.
>
>```
>system_default = tls_system_default
>
>[tls_system_default]
>MaxProtocol = TLSv1.2
>MaxProtocol = DTLSv1.2
>```
>
>In such a case we should not assume that the system default TLS settings should apply to QUIC. It makes no sense to do so.

Huh? It makes sense to do so to me. No reason not to apply system default TLS settings.

>In my view the former is the way to go and the latter introduces unnecessary differences in the API for QUIC vs TLS (i.e. you are forced to configure the allowed protocol versions differently between QUIC and TLS which goes against what we have been trying to do).

I have the opposite view. There should be one function for configuring the TLS version in use regardless of whether QUIC is in use. Anything else is just a perverse annoyance for application developers.

**Comment by hlandau** [comment]
> > Huh? It makes sense to do so to me. No reason not to apply system default TLS settings.
> 
> For min/max protocol? Why? Unless that min/max protocol happens to include TLSv1.3 it will not work for QUIC (in 3.2). I don't think it is at all unusual to set a max protocol version of TLSv1.2 in the system config. Such a setting might make perfect sense for normal TLS but is nonsensical for QUIC.

If a site has decided to configure a global TLS MaxProtocol value for TLS 1.2, it's likely that there is some administrative motivation for this. In this case the intersection of the set of versions which QUIC can use, and the set of versions administratively permitted, is the empty set; thus it follows that QUIC can't be used and attempts to use QUIC will fail. In my view this is how things should work.

Of course we could add a `QuicMaxVersion` option to the config processor that lets a different value be configured for QUIC connections, if site administrators want for some reason to allow TLS 1.3 only for QUIC connections.

**Comment by mattcaswell** [comment]
> I strongly disagree with the claim that because TLS 1.4 doesn't exist, or because QUICv2 isn't supported by us yet, that we should somehow defer these issues. 

All I am saying we should defer is the problem of what the API looks like for setting the underlying TLS version. Whichever route you go you are deferring something, i.e. if you go with set_min/max_protocol always means TLS version then you are deferring the problem of how to set the QUIC version. I don't object to defining that API now if you really want to - but it seems unnecessary to me.

**Comment by mattcaswell** [comment]
> If a site has decided to configure a global TLS MaxProtocol value for TLS 1.2, it's likely that there is some administrative motivation for this. In this case the intersection of the set of versions which QUIC can use, and the set of versions administratively permitted, is the empty set; thus it follows that QUIC can't be used and attempts to use QUIC will fail. In my view this is how things should work.

This sounds like a recipe for confusion and misconfiguration to me. I predict lots of problems if we go that route.

**Comment by hlandau** [comment]
> > I strongly disagree with the claim that because TLS 1.4 doesn't exist, or because QUICv2 isn't supported by us yet, that we should somehow defer these issues.
> 
> All I am saying we should defer is the problem of what the API looks like for setting the underlying TLS version. Whichever route you go you are deferring something, i.e. if you go with set_min/max_protocol always means TLS version then you are deferring the problem of how to set the QUIC version. I don't object to defining that API now if you really want to - but it seems unnecessary to me.

We already have an API for setting TLS versions... but set that aside for a moment, how is the SSL_set_min_protocol_version API supposed to work with QUIC version numbers? QUICv1 is 0x1, QUICv2 is 0x6b3343cf, QUICv3 could be e.g. 0x4444. This isn't a monotonically increasing sequence. Nor is it a valid assumption that if someone is okay with using QUICv(n+1) that they're also OK with using QUICv(n).

i.e., just figuring out how to map QUIC versions to the existing API is a significant design question and it doesn't seem to me that the existing API is suitable.

So the options are:
- Use `SSL_set_min_protocol_version` for  TLS versions, and defer QUIC version API design work (and  when do design that new API, can design it to the requirements of QUIC)
- Use `SSL_set_min_protocol_version` for QUIC versions, and deal with the QUIC version mapping conundrum *now* (but it is unclear to me it can be made to fit into this API); this does potentially save the design of a TLS version API until later but this is not a significant saving in design work as it would likely just be a copy/paste of the existing `SSL_set_min_protocol_version`-style API, just renamed

**Comment by mattcaswell** [comment]
> We already have an API for setting TLS versions... but set that aside for a moment, how is the SSL_set_min_protocol_version API supposed to work with QUIC version numbers? QUICv1 is 0x1, QUICv2 is 0x6b3343cf, QUICv3 could be e.g. 0x4444. This isn't a monotonically increasing sequence. 

I don't see how that has an impact on anything. We don't care what the actual version numbers used on the wire are. All we need to know for min/max protocol is that *logically* QUICv1 < QUICv2 < QUICv3. It doesn't have to be mathematically true for it to work (indeed logically DTLSv1 < DTLSv1.2 as far as min/max protocol is concerned, but in terms of the actual on-the-wire protocol version numbers DTLSv1.2 < DTLSv1).

**Comment by hlandau** [comment]
The existing API takes an `int`. QUIC version numbers are a 32-bit space. It's entirely possible `0x0303` will be a valid QUIC version number in future.

Moreover the more significant issue is that it's not a safe assumption that QUIC versions will be released in some ascending sequence of desirability. What if I want to support QUIC versions 1, 3, 5, 7 but not 2, 4, and 6? This isn't really an idle concern with the QUIC version registry being open to registration. Fundamentally this needs to be a list rather than two integers defining a range.

**Comment by mattcaswell** [comment]
> It's entirely possible 0x0303 will be a valid QUIC version number in future.

That is technically a problem...but it seems unlikely. Actually there's no reason that the version numbers we use for the API actually correspond to real on the wire protocol version numbers (they just always have been). They could be anything - so there is a workaround in the unlikely event that that occurs. I think its more important that we make the fundamental decision about whether set_min/max_protocol is about setting the version number for the primary protocol, or the underlying TLS version.

> What if I want to support QUIC versions 1, 3, 5, 7 but not 2, 4, and 6? 

We had that flexibility in the past for TLS versions - but we discouraged it favour of the min/max protocol approach because it turned out to be problematic and undesirable. I wouldn't recommend supporting that for QUIC either.

**Comment by hlandau** [comment]
>We had that flexibility in the past for TLS versions - but we discouraged it favour of the min/max protocol approach because it turned out to be problematic and undesirable. I wouldn't recommend supporting that for QUIC either.

The problem is that QUIC is already driving this sort of requirement. For example, QUICv2 includes no real functional changes or enhancements to QUICv1. Literally the only change in QUICv2 is a different version number and a different Initial packet key. The object of QUICv2 is for experimentation with how well QUIC's anti-ossification design aspects are working. Suppose QUICv3 comes out with real enhancements and someone wants to use it. It doesn't necessarily follow that they want to support QUICv2 also.

In other words, a range is I think a reasonable design for TLS but much less so for QUIC.

**Comment by hlandau** [comment]
> > It doesn't necessarily follow that they want to support QUICv2 also.
> 
> I cannot see why we should allow disabling just such QUICv2 if the application wants to support QUICv1 and QUICv3. It costs virtually nothing to support such QUICv2 in addition to supporting these two other versions and apart from some purely-for-testing purposes I do not see any reasonable use case.

There is a potential cost to supporting a given version, especially if load balancers or middleboxes are involved.

But in any case it doesn't follow that there is some total ordering on the set of QUIC versions by desirability. 'QUICv(n+1) is more desirable than QUICv(n)' cannot in general be assumed to be true. The use of numbers here is misleading; it would be more appropriate to say "is QUIC Red better than QUIC Blue? Should someone who wants QUIC Green also have to opt into QUIC Red?"

**Comment by hlandau** [comment]
Of course, a call to set a max version of TLS 1.2 could be silently clamped to set a max version of TLS 1.3. That would also be valid approach.

Can you articulate why `SSL_set_min_proto_version` shouldn't forward to the handshake layer? All existing use of the function is to configure the handshake layer. No application has ever existed that's used it to do anything else.

**Comment by mattcaswell** [comment]
> In that case I'd vote for simply making the set_min/max_proto to be ignored on QUIC contexts and SSL objects and the get_min/max_ always returning 0. 

I am not at all convinced by the red/blue argument. But I could live with this compromise.

**Comment by mattcaswell** [comment]
> I would be willing to accept SSL_set_min_proto_version not doing anything on a QUIC SSL object, (presumably with us defining an identikit copy called SSL_set_min_handshake_proto_version whenever TLS 1.4 comes out) and a new API for QUIC version management when we implement another QUIC version.

I think its more about deferring the whole decision to some future OpenSSL version. My agreement is on the basis that it doesn't do anything *right now* - but at some later point we might want to reopen the whole debate and make it do something.

**Comment by mattcaswell** [comment]
Ummm....although, at the risk of reopening the debate, it occurs to me that we would then be unable to redefine min/max_proto_version to have an impact on the underlying TLS handshake in OpenSSL 3.3 or above because it wouldn't be backwards compatible at that point.

E.g. if someone uses this config file in 3.2 and is doing QUIC:

````
system_default = tls_system_default

[tls_system_default]
MaxProtocol = TLSv1.2
MaxProtocol = DTLSv1.2
`````
Then that has no impact on QUIC with the compromise solution. If we later decide to adopt Hugo's interpretation of how this should work in 3.3 or later, then suddenly existing QUIC deployments would be broken.

**Comment by hlandau** [comment]
If we did assign semantics to `SSL_set_max_proto_version` at a later date, we could define a `QuicMaxHandshakeProtocol` property in config files, and ignore MaxProtocol if we're using a QUIC SSL object. I'd be OK with that. The sense I'm getting from both @mattcaswell and @t8m in any case is that MaxProtocol=TLSv1.2 shouldn't preclude use of QUIC, so for consistency we probably shouldn't apply it to QUIC objects at all.

**Comment by t8m** [comment]
> Ummm....although, at the risk of reopening the debate, it occurs to me that we would then be unable to redefine min/max_proto_version to have an impact on the underlying TLS handshake in OpenSSL 3.3 or above because it wouldn't be backwards compatible at that point.

It would be backwards compatible if we keep treating all versions <TLS-1.3 ignored on QUIC objects. If we start supporting TLS-1.4 and we make it possible by these calls to select TLS-1.3 or TLS-1.4 on the handshake layer, treating anything lower as TLS-1.3, it won't be backwards incompatible.
